src/gas/GasFeeController.ts

```diff
@@ -522,6 +579,69 @@ export class GasFeeController extends BaseController<
       this.state.gasFeeEstimates,
     );
   }
+
+  private async startPolling() {
+    if (this.intervalId === undefined) {
+      this.scheduleNextPoll();
+    }
+  }
+
+  private scheduleNextPoll() {
+    this.intervalId = setTimeout(async () => {
+      await this.poll();
+      this.scheduleNextPoll();
+    }, this.intervalDelay);
+  }
+
+  /**
+   * Removes the given item from the polling queue such that on the next
+   * iteration of the poll the corresponding request for the item will no longer
+   * be made.
+   *
+   * @param item - The known name of a piece of data for which we are polling.
+   */
+  stopPollingFor(item: keyof PollableItems) {
+    this.pollQueue.delete(item);
+  }
+
+  private pausePolling() {
+    if (this.intervalId !== undefined) {
+      clearInterval(this.intervalId);
+      this.intervalId = undefined;
+    }
+  }
+
+  private async determineWhetherNetworkCongested(): Promise<boolean> {
+    const isEIP1559Compatible = await this.getEIP1559Compatibility();
+    let isNetworkCongested = false;
+
+    if (isEIP1559Compatible) {
+      const feeHistory = await this.fetchFeeHistory({
+        ethQuery: this.ethQuery,
+        numberOfBlocks: 100,
+        percentiles: [50],
+      });
+      const sortedPriorityFees: number[] = feeHistory.blocks
+        .map((block) => block.priorityFeesByPercentile.get(50))
+        .filter(isNumber)
+        .sort((a, b) => a - b);
+
+      if (sortedPriorityFees.length > 0) {
+        const minPriorityFee = sortedPriorityFees[0];
+        const maxPriorityFee =
+          sortedPriorityFees[sortedPriorityFees.length - 1];
+        isNetworkCongested = maxPriorityFee >= minPriorityFee * 1.1;
```

Comment:
This is the algorithm I chose to determine whether the network is congested, but please note that I haven't run this by anything yet — I just went with something to get this PR started. That's why this is still a draft. There are ongoing discussions about this in Slack, so stay tuned.

Reply:
Most recent discussion is here: https://consensys.slack.com/archives/C021UN08FBJ/p1632245722147100